core/src/main/java/com/cryptomorin/xseries/XEntity.java

```diff
@@ -269,13 +293,76 @@ public static Entity spawn(@NotNull Location location, @NotNull ConfigurationSec
         XEntityType finalType = type.get().or(XEntityType.ZOMBIE);
         if (!finalType.isSupported()) return null;
 
+        return spawn(location, finalType.get().getEntityClass(), false, entity -> edit(entity, config));
+    }
+
+    @NotNull
+    @SuppressWarnings("unchecked")
+    public static <T extends Entity> T spawn(@NotNull Location location, @NotNull Class<T> clazz, boolean randomizeData, @Nullable Consumer<? super T> function) {
         if (SUPPORTS_DELAYED_SPAWN) {
-            return location.getWorld().spawn(location, finalType.get().getEntityClass(), false, entity -> edit(entity, config));
+            return location.getWorld().spawn(location, clazz, randomizeData, function);
+        } else if (DELAYED_SPAWN_1_17 != null) {
+            try {
+                return (T) DELAYED_SPAWN_1_17.invoke(location.getWorld(), location, clazz, randomizeData, oldConsumer(function));
+            } catch (Throwable t) {
+                throw new RuntimeException(t);
+            }
+        } else if (DELAYED_SPAWN_1_11 != null) {
+            try {
+                return (T) DELAYED_SPAWN_1_11.invoke(location.getWorld(), location, clazz, oldConsumer(function));
+            } catch (Throwable t) {
+                throw new RuntimeException(t);
+            }
         } else {
-            return edit(location.getWorld().spawnEntity(location, finalType.get()), config);
+            return location.getWorld().spawn(location, clazz);
```

Comment:
Consumer is not applied on the spawned entity. This leads to inconsistent behavior.